### PR TITLE
[ADP-3215] Add `TxId` type to `Cardano.Wallet.Read`

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
This pull request adds a type `TxId` to the `Read` hierarchy, and also starts to split the hierarchy into two parts:

* `Cardano.Read.Ledger` — presents an era-indexed interface to the ledger types
* `Cardano.Wallet.Read` — uses the above types to present a self-contained, era-indexed and hopefully simple view of the ledger concepts.

### Comments

On `Cardano.Read.Ledger`:

* I have refrained from renaming the existing modules in `Cardano.Wallet.Read` — I intend to move them later.

On `Cardano.Wallet.Read`:

* The `TxId` type is era-_independent_: A `TxId` that was parsed in one `era` may refer to a transaction which is from an entirely different era, and we can't even statically tell from which one. Hence, the type has no `era`-parameter.
* For the internal representation of `TxId`, I have chosen to be compatible with the type `TxId` from the shelley-style ledgers with zero-cost conversion. For Byron-transactions we do need to convert, though.
* This pull request prepares the addition of a `TxIn` type.

### Issue Number

ADP-3215

### CODE REVIEW

File: lib/read/lib/Cardano/Wallet/Read/Tx/TxId.hs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    Shelley -> L.unTxId
+    Allegra -> L.unTxId
+    Mary -> L.unTxId
+    Alonzo -> L.unTxId
+    Babbage -> L.unTxId
+    Conway -> L.unTxId
+
+fromByronTxId :: BY.Hash a -> TxId
+fromByronTxId =
+    txIdFromHash . fromJust . hashFromBytes . BY.hashToBytes
```
Comment (Anviking) [inline lib/read/lib/Cardano/Wallet/Read/Tx/TxId.hs]: Could also use `Cardano.Ledger.Shelley.API.ByronTranslation.translateTxIdByronToShelley` if you want to make it more obvious that this is safe

Comment (HeinrichApfelmus) [reply lib/read/lib/Cardano/Wallet/Read/Tx/TxId.hs]: That is an excellent idea! Done. Thank you! 😊

File: lib/read/test/Cardano/Wallet/Read/Tx/TxIdSpec.hs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+unsafeParseEraTxFromHex
+    :: forall era. IsEra era
+    => ByteString
+    -> Tx era
+unsafeParseEraTxFromHex bytes =
+    unsafeFromEraValue
+    . either (error . show) id
+    . parseTxFromCBOR
+    $ EraValue (K $ unsafeReadBase16 bytes :: K BL.ByteString era)
```
Comment (Anviking) [inline lib/read/test/Cardano/Wallet/Read/Tx/TxIdSpec.hs]: 🧐 `fromEraValue` seems sketchy to me. Should it really exist? What if we here instead wrote:

```suggestion
    either (error . show) id
    $ unComp
    $ deserializeTx
    (K $ unsafeReadBase16 bytes :: K BL.ByteString era)
  where
    unComp :: (f :.: g) era -> f (g era)
    unComp (Comp fg) = fg
```

**Review by Anviking — APPROVED** [review]
LGTM modulo https://github.com/cardano-foundation/cardano-wallet/pull/4619/files#r1692748006

File: lib/read/test/Cardano/Wallet/Read/Tx/TxIdSpec.hs (hunk 2 above)
Comment (HeinrichApfelmus) [reply lib/read/test/Cardano/Wallet/Read/Tx/TxIdSpec.hs]: The function `fromEraValue` is no more fishy than the function `read :: Read a => String → a` — this function is not parametrically polymorphic, the result value depends heavily on the result type.

A better name for the function `fromEraValue` would probably be `matchEraValue`, though.

But sure, I can remove this function.

Comment (Anviking) [reply lib/read/test/Cardano/Wallet/Read/Tx/TxIdSpec.hs]: > A better name for the function fromEraValue would probably be matchEraValue, though.

👍

**Pull request merged**